src/main/java/com/avispl/symphony/dal/communicator/aver/ptz/UDPCommunicator.java

```diff
@@ -0,0 +1,493 @@
+					this.status.setConnectionState(ConnectionState.Connected);
+					this.status.setLastError(null);
+				}
+			} catch (Exception ex) {
+				if (this.logger.isErrorEnabled()) {
+					this.logger.error("Error connecting to: " + this.host + " port: " + this.port, ex);
+				}
+
+				this.status.setLastError(ex);
+				this.status.setConnectionState(ConnectionState.Failed);
+				this.destroyChannel();
+				throw ex;
+			} finally {
+				writeLock.unlock();
+			}
+		}
+	}
+
+	/**
+	 * This method is used to disconnect from the device
+	 */
+	@Override
+	public void disconnect() {
+		if (this.logger.isTraceEnabled()) {
+			this.logger.trace("Disconnecting from: " + this.host + " port: " + this.port);
+		}
+
+		Lock writeLock = this.lock.writeLock();
+		writeLock.lock();
+
+		try {
+			this.destroyChannel();
+			this.status.setConnectionState(ConnectionState.Disconnected);
+		} finally {
+			writeLock.unlock();
+		}
+	}
+
+	/**
+	 * This method is used to get current connection status from the device
+	 */
+	@Override
+	public ConnectionStatus getConnectionStatus() {
+		Lock readLock = this.lock.readLock();
+		readLock.lock();
+
+		ConnectionStatus status;
+		try {
+			status = this.status.copyOf();
+		} finally {
+			readLock.unlock();
+		}
+
+		return status;
+	}
+
+	/**
+	 * This method is used to create a channel actually create a socket
+	 */
+	private void createChannel() {
+		try {
+			if (this.datagramSocket == null || this.datagramSocket.isClosed() || !this.datagramSocket.isConnected()) {
+				this.address = InetAddress.getByName(this.host);
+				this.datagramSocket = new DatagramSocket();
+				this.datagramSocket.setSoTimeout(this.timeout);
+			}
+
+		} catch (IOException ex) {
+			if (this.logger.isErrorEnabled()) {
+				this.logger.error("Error create UDP socket channel", ex);
+			}
+		}
+	}
+
+	/**
+	 * This method is used to destroy a channel actually destroy a socket
+	 */
+	public void destroyChannel() {
+		if (null != this.datagramSocket) {
+			try {
+				if (this.datagramSocket.isConnected()) {
+					this.datagramSocket.close();
+				}
+			} catch (Exception ex) {
+				if (this.logger.isWarnEnabled()) {
+					this.logger.warn("error seen on destroyChannel", ex);
+				}
+			}
+
+			this.datagramSocket = null;
+		}
+
+	}
+
+	/**
+	 * This method is used to check if a channel is connected or not
+	 */
+	private boolean isChannelConnected() {
+		return null != this.datagramSocket && this.datagramSocket.isConnected();
+	}
+
+	/**
+	 * This method is used to send a command to a device
+	 *
+	 * @param data This is the data to be sent
+	 * @return byte[] This returns the reply received from the device.
+	 */
+	protected byte[] send(byte[] data) throws Exception {
+		if (!this.isInitialized()) {
+			throw new IllegalStateException("ShellCommunicator cannot be used before init() is called");
+		} else if (null == data) {
+			throw new IllegalArgumentException("Send data is null");
+		} else {
```

Comment:
The exception will be thrown in if block, I think we can remove 2 `else` here. I will be more readable

Reply:
ok I'll update it